Source/Diagnostics/ReducedDiags/ParticleHistogram.cpp

```diff
@@ -0,0 +1,171 @@
+    }
+
+    // resize data array
+    m_data.resize(m_bin_num,0.0);
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs;
+            ofs.open(m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app);
+            // write header row
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            for (int i = 0; i < m_bin_num; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" + std::to_string(3+i) + "]";
+                Real b = m_bin_min + m_bin_size*(Real(i)+0.5);
+                ofs << "bin" + std::to_string(1+i)
+                             + "=" + std::to_string(b) + "()";
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+
+}
+// end constructor
+
+// function that computes the histogram
+void ParticleHistogram::ComputeDiags (int step)
+{
+
+    // Judge if the diags should be done
+    if ( (step+1) % m_freq != 0 ) return;
+
+    // get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    // get time at level 0
+    auto const t = warpx.gett_new(0);
+
+    // get MultiParticleContainer class object
+    auto & mypc = warpx.GetPartContainer();
+
+    // get WarpXParticleContainer class object
+    auto const & myspc = mypc.GetParticleContainer(m_selected_species_id);
+
+    using PType = typename WarpXParticleContainer::SuperParticleType;
+
+    // get parser
+    ParserWrapper<7> *fun_partparser = m_parser.get();
+
+    // declare local variables
+    Real const bin_min  = m_bin_min;
+    Real const bin_size = m_bin_size;
+    bool is_unity_particle_weight = false;
+    if ( m_norm == "unity_particle_weight" ) is_unity_particle_weight = true;
+
+    for ( int i = 0; i < m_bin_num; ++i )
+    {
+        // compute the histogram
+        m_data[i] = ReduceSum( myspc,
+        [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
+        {
+            auto const w  = p.rdata(PIdx::w);
+            auto const x  = p.pos(0);
+            auto const y  = p.pos(1);
+            auto const z  = p.pos(2);
+            auto const ux = p.rdata(PIdx::ux)/PhysConst::c;
+            auto const uy = p.rdata(PIdx::uy)/PhysConst::c;
+            auto const uz = p.rdata(PIdx::uz)/PhysConst::c;
+            auto const f = (*fun_partparser)(t,x,y,z,ux,uy,uz);
+            auto const f1 = bin_min + bin_size*i;
+            auto const f2 = bin_min + bin_size*(i+1);
+            if ( f > f1 && f < f2 ) {
+                if ( is_unity_particle_weight ) return 1.0;
+                else return w;
+            } else return 0.0;
+        });
+        // reduced sum over mpi ranks
+        ParallelDescriptor::ReduceRealSum
+            (m_data[i], ParallelDescriptor::IOProcessorNumber());
```

Comment:
This means one MPI-reduce per bin, I am afraid that this will not be acceptable performance-wise. could you put this reduction outside of the `for ( int i = 0; i < m_bin_num; ++I )` loop to have a single reduction in this function?

Reply:
I tried to put it outside the `for` loop like:
```
ParallelDescriptor::ReduceRealSum
            (m_data, ParallelDescriptor::IOProcessorNumber());
```
but it seems like a vector like `m_data` does not work, it may has to be a single double variable that passed into the function `ParallelDescriptor::ReduceRealSum`.

Reply:
@atmyers Ideally this would be a `ReduceSum` on a `std::vector<amrex::Real>` (where the vector will typically have 10-1000 elements). Is this implemented in AMReX? Otherwise, what would you suggest to do this communication?

Reply:
@MaxThevenet @atmyers I saw similar things in `amrex/Tools/C_util/Convergence/ComputeAmrDataNorms.cpp`
```
for (int iComp=0; iComp<nComp; ++iComp)
	    {
		ParallelDescriptor::ReduceRealMax(n0[iComp]);
		ParallelDescriptor::ReduceRealSum(n1[iComp]);
		ParallelDescriptor::ReduceRealSum(n2[iComp]);
		
		norm0[iComp] = std::max(norm0[iComp],n0[iComp]);
		norm1[iComp] += n1[iComp];
		norm2[iComp] += n2[iComp];
	    }
```
Can we simply keep the old way for now?
```
ParallelDescriptor::ReduceRealSum
            (m_data[i], ParallelDescriptor::IOProcessorNumber());
```